### ISSUE TITLE
11 test failures with PyPy3.9 7.3.11 in 2.0.0 (and git main)

### ISSUE THREAD

**Issue by mgorny** [sqlalchemy/sqlalchemy#9154]
### Describe the bug

When running the test suite (as of 6779fa861d11668f67d780d77e4ae76965cee6f7) under PyPy3.9 7.3.11, I get the following test failures:

```
AILED test/ext/test_associationproxy.py::DictOfTupleUpdateTest::test_update_multi_elem_varg - AssertionError: 'update expected at most 1 arguments?, got 2' !~ dict.update() takes from 0 to 1 positional arguments but 2 were g...
FAILED test/ext/test_associationproxy.py::DictOfTupleUpdateTest::test_update_one_elem_varg - AssertionError: 'dictionary update sequence element #1 has length 5; 2 is required' !~ sequence of pairs expected
FAILED test/ext/test_associationproxy.py::ProxyHybridTest::test_msg_fails_on_cls_access - AssertionError: 'Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from c...
FAILED test/dialect/test_sqlite.py::TestTypes_sqlite+pysqlite_3_40_1::test_cant_parse_datetime_message - AssertionError: Exception ValueError("Invalid isoformat string: 'ASDF'") was correctly raised but did not set a cause, within cont...
FAILED test/engine/test_processors.py::PyDateProcessorTest::test_date_invalid_string - AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within conte...
FAILED test/engine/test_processors.py::PyDateProcessorTest::test_datetime_invalid_string - AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within conte...
FAILED test/engine/test_processors.py::PyDateProcessorTest::test_time_invalid_string - AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within conte...
FAILED test/dialect/test_suite.py::ReturningGuardsTest_sqlite+pysqlite_3_40_1::test_delete_single - AssertionError: Callable did not raise an exception
FAILED test/dialect/test_suite.py::ReturningGuardsTest_sqlite+pysqlite_3_40_1::test_insert_single - AssertionError: Callable did not raise an exception
FAILED test/dialect/test_suite.py::ReturningGuardsTest_sqlite+pysqlite_3_40_1::test_update_single - AssertionError: Callable did not raise an exception
FAILED test/base/test_utils.py::ImmutableDictTest::test_pep584 - AttributeError: 'super' object has no attribute '__ror__'. Did you mean: '__dir__'?
```

Some of them might be valid bugs in PyPy; I can pass the report on to PyPy developers if you wish.

### To Reproduce

```python
tox -e pypy3  # ;-)
```

### Error

<details>
<summary>Full tracebacks</summary>

```pytb
============================================================== FAILURES ===============================================================
__________________________________________ DictOfTupleUpdateTest.test_update_multi_elem_varg __________________________________________
[gw0] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 1755, in update
    up.update(*a, **kw)
TypeError: dict.update() takes from 0 to 1 positional arguments but 2 were given

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/ext/test_associationproxy.py", line 2404, in test_update_multi_elem_varg
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 450, in _expect_raises
    assert re.search(msg, error_as_string, re.UNICODE), "%r !~ %s" % (
AssertionError: 'update expected at most 1 arguments?, got 2' !~ dict.update() takes from 0 to 1 positional arguments but 2 were given
assert None
 +  where None = <function search at 0x00007fd3de58d880>('update expected at most 1 arguments?, got 2', 'dict.update() takes from 0 to 1 positional arguments but 2 were given', re.UNICODE)
 +    where <function search at 0x00007fd3de58d880> = re.search
 +    and   re.UNICODE = re.UNICODE
------------------------------------------------------ Captured stderr teardown -------------------------------------------------------
Error emptying table b: OperationalError('(_sqlite3.OperationalError) no such table: b')
Error emptying table a: OperationalError('(_sqlite3.OperationalError) no such table: a')
___________________________________________ DictOfTupleUpdateTest.test_update_one_elem_varg ___________________________________________
[gw0] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 1755, in update
    up.update(*a, **kw)
ValueError: sequence of pairs expected

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/ext/test_associationproxy.py", line 2394, in test_update_one_elem_varg
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 450, in _expect_raises
    assert re.search(msg, error_as_string, re.UNICODE), "%r !~ %s" % (
AssertionError: 'dictionary update sequence element #1 has length 5; 2 is required' !~ sequence of pairs expected
assert None
 +  where None = <function search at 0x00007fd3de58d880>('dictionary update sequence element #1 has length 5; 2 is required', 'sequence of pairs expected', re.UNICODE)
 +    where <function search at 0x00007fd3de58d880> = re.search
 +    and   re.UNICODE = re.UNICODE
------------------------------------------------------ Captured stderr teardown -------------------------------------------------------
Error emptying table b: OperationalError('(_sqlite3.OperationalError) no such table: b')
Error emptying table a: OperationalError('(_sqlite3.OperationalError) no such table: a')
____________________________________________ ProxyHybridTest.test_msg_fails_on_cls_access _____________________________________________
[gw0] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 639, in for_proxy
    cls._cls_unwrap_target_assoc_proxy(target_class, value_attr),
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 710, in _cls_unwrap_target_assoc_proxy
    attr = getattr(target_class, value_attr)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/hybrid.py", line 932, in __get__
    return self._expr_comparator(owner)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/hybrid.py", line 1162, in expr_comparator
    comparator(owner),
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/hybrid.py", line 1132, in _expr
    return ExprComparator(cls, expr(cls), self)
  File "/tmp/sqlalchemy/test/ext/test_associationproxy.py", line 3424, in fails_on_class_access
    return len(self.data)
TypeError: 'InstrumentedAttribute' has no length

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/test/ext/test_associationproxy.py", line 3445, in test_msg_fails_on_cls_access
    a1.fails_on_class_access
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 426, in __get__
    inst = self._as_instance(owner, instance)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 491, in _as_instance
    inst = AssociationProxyInstance.for_proxy(self, owner, obj)
  File "/tmp/sqlalchemy/lib/sqlalchemy/ext/associationproxy.py", line 649, in for_proxy
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from class "B": 'InstrumentedAttribute' has no length

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/ext/test_associationproxy.py", line 3445, in test_msg_fails_on_cls_access
    a1.fails_on_class_access
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 450, in _expect_raises
    assert re.search(msg, error_as_string, re.UNICODE), "%r !~ %s" % (
AssertionError: 'Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from class "B": .* no len\\(\\)' !~ Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from class "B": 'InstrumentedAttribute' has no length
assert None
 +  where None = <function search at 0x00007fd3de58d880>('Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from class "B": .* no len\\(\\)', 'Association proxy received an unexpected error when trying to retreive attribute "B.fails_on_class_access" from class "B": \'InstrumentedAttribute\' has no length', re.UNICODE)
 +    where <function search at 0x00007fd3de58d880> = re.search
 +    and   re.UNICODE = re.UNICODE
------------------------------------------------------ Captured stderr teardown -------------------------------------------------------
Error emptying table c: OperationalError('(_sqlite3.OperationalError) no such table: c')
Error emptying table b: OperationalError('(_sqlite3.OperationalError) no such table: b')
Error emptying table a: OperationalError('(_sqlite3.OperationalError) no such table: a')
__________________________________ TestTypes_sqlite+pysqlite_3_40_1.test_cant_parse_datetime_message __________________________________
[gw2] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/usr/lib/pypy3.9/datetime.py", line 1488, in fromisoformat
    return cls(*_parse_isoformat_time(time_string))
  File "/usr/lib/pypy3.9/datetime.py", line 335, in _parse_isoformat_time
    time_comps = _parse_hh_mm_ss_ff(timestr)
  File "/usr/lib/pypy3.9/datetime.py", line 296, in _parse_hh_mm_ss_ff
    time_comps[comp] = int(tstr[pos:pos+2])
ValueError: invalid literal for int() with base 10: 'AS'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/test/dialect/test_sqlite.py", line 122, in <lambda>
    lambda: connection.execute(
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/result.py", line 1541, in scalar
    return self._only_one_row(
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/result.py", line 749, in _only_one_row
    row = make_row(row) if make_row else row
  File "/usr/lib/pypy3.9/functools.py", line 302, in __call__
    return self.func(*self.args, *args, **keywords)
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_row.py", line 64, in __init__
    [
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_row.py", line 65, in <listcomp>
    proc(value) if proc else value
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_processors.py", line 125, in str_to_time
    dt_value = time_cls.fromisoformat(value)
  File "/usr/lib/pypy3.9/datetime.py", line 1490, in fromisoformat
    raise ValueError(f'Invalid isoformat string: {time_string!r}')
ValueError: Invalid isoformat string: 'ASDF'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/dialect/test_sqlite.py", line 119, in test_cant_parse_datetime_message
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 455, in _expect_raises
    _assert_proper_exception_context(err)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 352, in _assert_proper_exception_context
    assert False, (
AssertionError: Exception ValueError("Invalid isoformat string: 'ASDF'") was correctly raised but did not set a cause, within context ValueError("invalid literal for int() with base 10: 'AS'") as its cause.
assert False
____________________________________________ PyDateProcessorTest.test_date_invalid_string _____________________________________________
[gw1] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/usr/lib/pypy3.9/datetime.py", line 895, in fromisoformat
    assert len(date_string) == 10
AssertionError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_processors.py", line 133, in str_to_date
    dt_value = date_cls.fromisoformat(value)
  File "/usr/lib/pypy3.9/datetime.py", line 898, in fromisoformat
    raise ValueError(f'Invalid isoformat string: {date_string!r}')
ValueError: Invalid isoformat string: '5:a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/engine/test_processors.py", line 105, in test_date_invalid_string
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 455, in _expect_raises
    _assert_proper_exception_context(err)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 352, in _assert_proper_exception_context
    assert False, (
AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within context AssertionError() as its cause.
assert False
__________________________________________ PyDateProcessorTest.test_datetime_invalid_string ___________________________________________
[gw1] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/usr/lib/pypy3.9/datetime.py", line 1781, in fromisoformat
    date_components = _parse_isoformat_date(dstr)
  File "/usr/lib/pypy3.9/datetime.py", line 273, in _parse_isoformat_date
    year = int(dtstr[0:4])
ValueError: invalid literal for int() with base 10: '5:a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_processors.py", line 117, in str_to_datetime
    dt_value = datetime_cls.fromisoformat(value)
  File "/usr/lib/pypy3.9/datetime.py", line 1783, in fromisoformat
    raise ValueError(f'Invalid isoformat string: {date_string!r}')
ValueError: Invalid isoformat string: '5:a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/engine/test_processors.py", line 113, in test_datetime_invalid_string
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 455, in _expect_raises
    _assert_proper_exception_context(err)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 352, in _assert_proper_exception_context
    assert False, (
AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within context ValueError("invalid literal for int() with base 10: '5:a'") as its cause.
assert False
____________________________________________ PyDateProcessorTest.test_time_invalid_string _____________________________________________
[gw1] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/usr/lib/pypy3.9/datetime.py", line 1488, in fromisoformat
    return cls(*_parse_isoformat_time(time_string))
  File "/usr/lib/pypy3.9/datetime.py", line 335, in _parse_isoformat_time
    time_comps = _parse_hh_mm_ss_ff(timestr)
  File "/usr/lib/pypy3.9/datetime.py", line 296, in _parse_hh_mm_ss_ff
    time_comps[comp] = int(tstr[pos:pos+2])
ValueError: invalid literal for int() with base 10: '5:'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 441, in _expect_raises
    yield ec
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/tmp/sqlalchemy/lib/sqlalchemy/engine/_py_processors.py", line 125, in str_to_time
    dt_value = time_cls.fromisoformat(value)
  File "/usr/lib/pypy3.9/datetime.py", line 1490, in fromisoformat
    raise ValueError(f'Invalid isoformat string: {time_string!r}')
ValueError: Invalid isoformat string: '5:a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/engine/test_processors.py", line 121, in test_time_invalid_string
    assert_raises_message(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 368, in assert_raises_message
    return _assert_raises(
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 418, in _assert_raises
    callable_(*args, **kwargs)
  File "/usr/lib/pypy3.9/contextlib.py", line 137, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 455, in _expect_raises
    _assert_proper_exception_context(err)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 352, in _assert_proper_exception_context
    assert False, (
AssertionError: Exception ValueError("Invalid isoformat string: '5:a'") was correctly raised but did not set a cause, within context ValueError("invalid literal for int() with base 10: '5:'") as its cause.
assert False
____________________________________ ReturningGuardsTest_sqlite+pysqlite_3_40_1.test_delete_single ____________________________________
[gw3] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 670, in test_delete_single
    run_stmt(stmt, False, "b_id", connection.dialect.delete_returning)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 595, in go
    connection.execute(
  File "/usr/lib/pypy3.9/contextlib.py", line 126, in __exit__
    next(self.gen)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 466, in _expect_raises
    assert success, "Callable did not raise an exception"
AssertionError: Callable did not raise an exception
assert False
____________________________________ ReturningGuardsTest_sqlite+pysqlite_3_40_1.test_insert_single ____________________________________
[gw3] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 611, in test_insert_single
    run_stmt(stmt, False, "id", connection.dialect.insert_returning)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 595, in go
    connection.execute(
  File "/usr/lib/pypy3.9/contextlib.py", line 126, in __exit__
    next(self.gen)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 466, in _expect_raises
    assert success, "Callable did not raise an exception"
AssertionError: Callable did not raise an exception
assert False
____________________________________ ReturningGuardsTest_sqlite+pysqlite_3_40_1.test_update_single ____________________________________
[gw3] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 636, in test_update_single
    run_stmt(stmt, False, "b_id", connection.dialect.update_returning)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/suite/test_dialect.py", line 595, in go
    connection.execute(
  File "/usr/lib/pypy3.9/contextlib.py", line 126, in __exit__
    next(self.gen)
  File "/tmp/sqlalchemy/lib/sqlalchemy/testing/assertions.py", line 466, in _expect_raises
    assert success, "Callable did not raise an exception"
AssertionError: Callable did not raise an exception
assert False
____________________________________________________ ImmutableDictTest.test_pep584 ____________________________________________________
[gw2] linux -- Python 3.9.16 /tmp/sqlalchemy/.tox/pypy3/bin/python
Traceback (most recent call last):
  File "/tmp/sqlalchemy/test/base/test_utils.py", line 356, in test_pep584
    i2 = {"x": 3} | i2
  File "/tmp/sqlalchemy/lib/sqlalchemy/util/_py_collections.py", line 145, in __ror__
    return immutabledict(super().__ror__(__value))
AttributeError: 'super' object has no attribute '__ror__'. Did you mean: '__dir__'?
-------------------------------------------------------- Captured stdout call ---------------------------------------------------------
b'immutabledict object is immutable'
``` 
</details>

### Versions

- OS: Gentoo Linux amd64
- Python: PyPy3.9 7.3.11 (with stdlib from `Python 3.9.16 (feeb267ead3e6771d3f2f49b83e1894839f64fb7, Dec 30 2022, 11:41:00)`)
- SQLAlchemy: 2.0.0, 6779fa861d11668f67d780d77e4ae76965cee6f7
- Database: n/a
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): n/a

**Comment by zzzeek** [comment]
were you running the 1.4 test suite under pypy?  at some point we had to take pypy out of the testing pipeline

**Comment by CaselIT** [comment]
We re-added it on GitHub, but we have set as optional so a fail does bot break the build.
I think a similar number of tests is breathing in the pipeline https://github.com/sqlalchemy/sqlalchemy/actions/runs/4019875533/jobs/6907218309#step:5:2018

**Comment by mgorny** [comment]
> were you running the 1.4 test suite under pypy? at some point we had to take pypy out of the testing pipeline

Yes. I have two marked as failing in 1.4.46. One of them is linked to #8762, so might have been fixed already. The other was an exception message mismatch that I probably didn't report.

Unfortunately, the test suite takes a long time and for some reason running it in parallel via xdist crashes immediately on me, so I have had limited opportunities to retest.

**Comment by zzzeek** [comment]
pypy's development velocity these days seems to be very low so it's difficult to keep the test suite lined up with it.  are you actually using pypy in production?

**Comment by mgorny** [comment]
I'm maintaining SQLAlchemy package for Gentoo, and our SQLAlchemy package installs for PyPy as well. I don't know if anyone's actually using it in production but it definitely has a number of reverse dependencies, so we have to continue providing support. I'm afraid that's the best answer I can give you because we have no way of measuring what people actually use (except for complaints when we remove stuff they happened to use ;-)).

**Comment by mattip** [comment]
Two thoughts:
- Is tox testing using coverage? PyPy + coverage + pytest is very slow. This is a known problem, and one we would love to solve but have not been able to so far. Could you exclude coverage when testing PyPy?
- I see there is an option to avoid the cython code [when installing](https://docs.sqlalchemy.org/en/20/intro.html#building-the-cython-extensions) via `export DISABLE_SQLALCHEMY_CEXT=1; python setup.py install`. If tests pass using this option, it might be faster on PyPy. In order to support the CPython C-API, PyPy needs to do a lot of emulation which slows it down. Also the JIT cannot peer into C code, so any opportunities to make things faster with the JIT are lost.

**Comment by CaselIT** [comment]
- in general no, tox doesn't run coverage
- the exclusion of cython should be automatic when installing, since it's active only on cpython https://github.com/sqlalchemy/sqlalchemy/blob/5f0c591c744ddf17cf5d886aec21e242eb58562f/setup.py#L33

**Comment by mattip** [comment]
Thanks for the clarifications and for considering PyPy in your project.

**Comment by CaselIT** [comment]
Just for reference, the ci on github runs tests on pypy, but it currently allows failures.
Results can be found in the actions table, example: https://github.com/sqlalchemy/sqlalchemy/actions/runs/5934122978/job/16090537958#step:6:2839

**Comment by mgorny** [comment]
By the way, the datetime exception bugs aren't really specific to PyPy — they also happen if you remove `_datetime` C ext from CPython, and it uses the pure Python implementation.

**Comment by CaselIT** [comment]
I haven't really looked at these errors yet, sorry

**Comment by zzzeek** [comment]
can we look at the conditions for this one?  just want to understand what was observed for this one:

```diff
diff --git a/lib/sqlalchemy/pool/impl.py b/lib/sqlalchemy/pool/impl.py
index 2807acf..307c4fc 100644
--- a/lib/sqlalchemy/pool/impl.py
+++ b/lib/sqlalchemy/pool/impl.py
@@ -442,7 +442,7 @@
         except AttributeError:
             pass
         else:
-            if rec is not None:
+            if rec is not None and not rec.is_detached:
                 return rec._checkout_existing()
 
         return _ConnectionFairy._checkout(self, self._fairy)
```

did a test fail or something?     looking at what that says, I would thing we should instead be saying `assert not rec.is_detached`.   is the issue that detachment is not addressing `current()` correctly / atomically?

**Comment by zzzeek** [comment]
actually this is singletonthreadpool which I dont think even supports detachment....

**Comment by mattip** [comment]
It fixed a pypy failure around released-but-not-gc-collected resources

**Comment by zzzeek** [comment]
was it like in the teardown of the whole suite?    or a specific test or suite?

**Comment by mattip** [comment]
I reverted the `not rec.is_detached` change, pending a deeper review of remaining PyPy failures after a new version of PyPy is released to fix some other incompatibilities.

**Comment by zzzeek** [comment]
OK...I mean i think the change is correct, it's just it implies someone actually called detach(), which is not a thing we do for SingletonThreadPool (just confirmed it), but looking at how "is_detached" works, it's not dependent on "detach()" being called, it could totally be triggered during a GC scenario as well.

looking at the code it looks like the error you would have gotten then would be the assertion error "can't 'checkout' a detached connection fairy" .

But why would pool.connect() be called within a gc scenario.  

basically it's a scenario that is unexpected so I usually try to get as much of an understanding of that scenario as possible to make sure we're addressing it in the best way.     but we'll go with omitting it for now.
